feat: Add metrics about subnet usage on the state tree (#191)
dfinity/interface-spec: spec/index.md

        principal = bytes .size (0..29)
        tagged<t> = #6.55799(t) ; the CBOR tag

> -   `/subnet/<subnet_id>/metrics` (blob)

Author: This is not fully fleshed out. We could also imagine an alternative of a separate path per metric. There's probably a trade-off here between having a single struct with all metrics and how difficult it is to verify the certificate for a single blob or having to perform verification for every different path listed.

        principal = bytes .size (0..29)
        tagged<t> = #6.55799(t) ; the CBOR tag

> -   `/subnet/<subnet_id>/metrics` (blob)

Reviewer: A danger of this endpoint is that replica-signed queries require read_state requests to `/subnet` and thus its responses would now also serve the metrics (not a problem is we don't have many metrics, but could be a problem in the future).

Author: I am not sure I'm following what's the problem. I suppose for replica-signed queries you don't strictly need the metrics but the metrics are also not a secret, so returning them seems ok security wise.

Reviewer 2: It is not about security but about data size. If the subnets contain all public keys of the nodes hosting the subnet, though, that that is a lot more data (~ 1KB per subnet) so the metrics at this point don't really add a lot of data on top of this.

Author: I see, thanks for the clarification. Yeah, I don't expect the metrics will ever be anything other than a few counters, so something in the order of hundreds of bytes extra in the worst case.

>      A list of subnet wide metrics related to this subnet's current usage and/or performance. The metrics are a CBOR map with the following fields:

Reviewer: ```suggestion
     A collection of subnet-wide metrics related to this subnet's current resource usage and/or performance. The metrics are a CBOR map with the following fields:
```
I'm using "collection" instead of "list" as the blob is a CBOR "map".

     
     - `num_canisters` (`nat`): The number of canisters on this subnet.

>      - `canister_state` (`nat`): The total state taken by canisters on this subnet in bytes.

Reviewer: ```suggestion
     - `canister_state` (`nat`): The total size of the state taken by canisters on this subnet in bytes.
```

>      - `consumed_cycles` (`nat`): The total number of cycles consumed by canisters on this subnet.

Author: These would be counters counting from when the subnet was created (not sure if that's what you meant by genesis here).

        principal = bytes .size (0..29)
        tagged<t> = #6.55799(t) ; the CBOR tag

-   `/subnet/<subnet_id>/metrics` (blob)

     A collection of subnet-wide metrics related to this subnet's current resource usage and/or performance. The metrics are a CBOR map with the following fields:
     
     - `num_canisters` (`nat`): The number of canisters on this subnet.
     - `total_canister_state` (`nat`): The total size of the state in bytes taken by canisters on this subnet since this subnet was created.
     - `total_consumed_cycles` (`nat`): The total number of cycles consumed by canisters on this subnet since this subnet was created.

>      - `num_update_transactions` (`nat`): The total number of transactions processed on this subnet since this subnet was created.

Reviewer 3: I would consider whether it makes sense to follow [Prometheus naming best practices](https://prometheus.io/docs/practices/naming/#metric-names) (or similar) when naming these metrics. Among other things, it's generally a good idea to include the measurement unit in the name (e.g. `total_canister_state_bytes` rather than a confusing `total_canister_state`). And it is good to identify counters, e,g, by giving them a `total` suffix (e.g. `total_consumed_cycles_total` or ``update_transactions_total`).

Author: I saw this late and the change has already been merged on the replica side. Given that these are not prometheus metrics I don't think we're bound to follow the best practices it establishes. We explain what the metrics represent so that should be ok. If spec owners feel strongly, I'm happy to change (although the hassle of changing some names given we're talking about certified state is likely not worth it imo).

Reviewer: I think Alin's suggestion to make the metric names more self-explanatory makes sense so I'd be up for it if it's still possible.

Author: Are the following changes ok with everyone?

`total_canister_state` -> `total_canister_state_bytes`
`num_update_transactions` -> `update_transactions_total`

The suggestion of `total_consumed_cycles_total` seems confusing, at least to me (there's two instances of "total" in the name).

Reviewer 3: I don't mind if we don't follow Prometheus naming practices specifically. The `total` can go at the end or at the start (or be replaced with something else, e.g. `all_time`). I would very much prefer it for the naming to be consistent, though. I.e. don't use `total` in both gauges and counters; and don't put it at the beginning of one metric name and the end of another.

One possible naming scheme (although, again, I'm not insisting it be followed to the letter) could be:
```
canisters
canister_state_bytes
consumed_cycles_total
update_transactions_total
```

Reviewer 3: One technical aspect of this (which David pointed out) is that since the metrics are encoded as a blob; and the canonical state uses packed CBOR encoding; the names can be changed at any time, without breaking compatibility.

E.g. `{"num_canisters": 5, "total_canister_state_bytes": 100}` would be encoded as `{0: 5, 1: 100}`, which is the same as what `{"canisters": 5, "canister_state_bytes": 100}` would encode to (as long as we preserved field ordering).

Author: Consumed cycles is a gauge and update transactions is a counter, so your suggestion is inconsistent with itself if we use `total` in both. If we are ok with that, then I am ok with everything else except `canisters`, I'd prefer to keep `num_canisters` as it's more descriptive imo.

Reviewer 3: Consumed cycles is a gauge only because it's not monotonically increasing (we subtract what are essentially reserved cycles and then return part of them later on). But leaving aside that awkwardness, it is essentially a counter, in that the absolute value is not particularly meaningful; what is meaningful about it by how much it changes over a day or an hour.

So it's all up in the air whether we want to call it a counter or a gauge. I would go with counter, since we are at very low risk of someone treating it strictly as a Prometheus counter and e.g. applying Prometheus' definition of `rate()` to it. People are just going to take the value now minus the value one hour ago. They may be surprised to see a negative difference, but that's about it.

Author: Alright, I buy this argument. I'll update to:

```
num_canisters
canister_state_bytes
consumed_cycles_total
update_transactions_total
```

Author: Updated.

>      - `num_update_transactions` (`nat`): The total number of transactions processed on this subnet since this subnet was created.

Reviewer 3: Not sure how transactions are counted, but consumed cycles are summed across all canisters (to which the cycles burned by deleted canisters are added). Currently, this is indeed "cycles consumed [...] since this subnet was created". But if we consider subnet splitting/canister migration, this no longer holds.

Maybe it makes sense to reword this (or maybe not). That being said, I don't have a suggestion for how to make this both generic and concise. Maybe "the sum of all-time burned cycles by all current and deleted canisters".

Author: @mraszyk any thoughts on this? I'm ok with "the sum of all-time burned cycles by all current and deleted canisters".

Re update_transactions: we keep accumulating them in a counter at the end of every round execution. So, it would include indeed all transactions since the subnet was created (even ones that were for canisters that were migrated away).

Reviewer: I'm also fine with the formulation "the sum of all-time burned cycles by all current and deleted canisters".

Author: Ok, changed it to the proposed formulation.

E.content = ReadState RS
TS = verify_envelope(E, RS.sender, S.system_time)
S.system_time <= RS.ingress_expiry

> ∀ path ∈ RS.paths. may_read_path_for_canister(S, R.sender, path)

Reviewer: ```suggestion
∀ path ∈ RS.paths. may_read_path_for_canister(S, RS.sender, path)
```
You didn't introduce this typo, but why not fix it?

-   `{certificate: C}`

The predicate `may_read_path_for_canister` is defined as follows, implementing the access control outlined in [Request: Read state](#http-read-state):

    may_read_path_for_canister(S, _, ["time"] · _) = False

>     may_read_path_for_canister(S, _, ["subnet"] · _) = False

Author: I believe we can actually drop these from here and allow implementations to serve these paths from both endpoints for backward compatibility. But I think it's also ok to have them be served by both endpoints. I'll go with whatever is the spec owner preference here.

Reviewer: If we define `False` here, then we it shouldn't be permitted to serve those paths (what's the point of this predicate if it was permitted anyway?).

Author: Ok, maybe I wasn't clear on my intention. I think it's fine from a spec point of view that we drop these from here (and therefore they would be caught in the last predicate below which is the catch all). This would mean that these paths should not be permitted given a new version of the spec released with this change. However, the replica implementation should not immediately drop support for those, otherwise we would have a breaking change (which I would like to avoid).

Reviewer: Discussed in the Interface spec meeting: we agreed to keep the `/time` and `/subnet` paths available via the `/api/v2/canister` URL, too (the `/time` is anyway always included and it makes sense to allow `/subnet` for backward-compatibility, too).

    may_read_path_for_canister(S, _, ["request_status", Rid] · _) =
      ∀ (R ↦ (_, ECID')) ∈ dom(S.requests). hash_of_map(R) = Rid => RS.sender == R.sender ∧ ECID == ECID'
    may_read_path_for_canister(S, _, ["canister", cid, "module_hash"] · _) = cid == ECID
    may_read_path_for_canister(S, _, ["canister", cid, "controllers"] · _) = cid == ECID
    may_read_path_for_canister(S, _, ["canister", cid, "metadata", name] · _) = cid == ECID ∧ UTF8(name) ∧
      (cid ∉ dom(S.canisters[cid]) ∨
       S.canisters[cid] = EmptyCanister ∨
       name ∉ (dom(S.canisters[cid].public_custom_sections) ∪ dom(S.canisters[cid].private_custom_sections)) ∨
       name ∈ dom(S.canisters[cid].public_custom_sections) ∨
       (name ∈ dom(S.canisters[cid].private_custom_sections) ∧ RS.sender ∈ S.controllers[cid])
      )
    may_read_path_for_canister(S, _, _) = False

where `UTF8(name)` holds if `name` is encoded in UTF-8.

Submitted request to `/api/v2/subnet/<subnet_id>/read_state`
`E`

Conditions  

```html

E.content = ReadState RS
TS = verify_envelope(E, RS.sender, S.system_time)
S.system_time <= RS.ingress_expiry

> ∀ path ∈ RS.paths. may_read_path_for_subnet(S, R.sender, path)

Reviewer: ```suggestion
∀ path ∈ RS.paths. may_read_path_for_subnet(S, RS.sender, path)
```
You didn't introduce this typo, but why not fix it?

```

Read response  
A record with

-   `{certificate: C}`
  

> The predicate `may_read_path_for_subnet` is defined as follows, implementing the access control outlined in [Request: Read state](#http-read-state):

Reviewer: Could you please refine the sentence "The subnet form should be used when the information to be retrieved is subnet specific" to hint at what paths are actually allowed in that case.

Author: Let me know what you think of the current version.

Reviewer: Thank you! Pushed a slight reformulation.

    may_read_path_for_subnet(S, _, ["time"] · _) = True
    may_read_path_for_subnet(S, _, ["subnet"] · _) = True
    may_read_path_for_subnet(S, _, _) = False

> where `UTF8(name)` holds if `name` is encoded in UTF-8.

Reviewer: not needed here

-   `{certificate: C}`

The predicate `may_read_path_for_canister` is defined as follows, implementing the access control outlined in [Request: Read state](#http-read-state):

    may_read_path_for_canister(S, _, ["time"]) = True
    may_read_path_for_canister(S, _, ["subnet"]) = True
    may_read_path_for_canister(S, _, ["subnet", sid]) = True
    may_read_path_for_canister(S, _, ["subnet", sid, "public_key"]) = True
    may_read_path_for_canister(S, _, ["subnet", sid, "canister_ranges"]) = True

>     may_read_path_for_canister(S, _, ["subnet", sid, "metrics"]) = True

Author: @mraszyk I would like to get a confirmation on what we are expecting to do with allowing these under the canister endpoint. In the implementation, I would definitely still allow them for backwards compatibility (and allow end users to transition to the new endpoint) but I think we could also already remove them from the spec already.

The metrics one seems safe to only allow in the new endpoint given that it's a new path on the tree, so at the very least this one can be removed from here.

On a separate note: I've seen in the code that we also allow some paths of the form `["subnet", sid, "node", node_id])`, is this some other PR still in progress, so it doesn't appear here?

Reviewer: > I would like to get a confirmation on what we are expecting to do with allowing these under the canister endpoint. In the implementation, I would definitely still allow them for backwards compatibility (and allow end users to transition to the new endpoint) but I think we could also already remove them from the spec as we are providing them through the new subnet endpoint.

I would keep them in the spec as long as we keep them in the actual implementation. If we want to deprecate them (by forbidding the paths in the HTTP handler), then I'd announce this intention as a note in the spec (at the moment we're just saying that the subnet form "should" be used and we can strengthen this statement in a further note).

> The metrics path seems safe to only allow in the new endpoint given that it's a new path on the tree, so at the very least this one can be removed from here.

Makes sense. Done

> On a separate note: I've seen in the code that we also allow some paths of the form ["subnet", sid, "node", node_id]), is this some other PR still in progress, so it doesn't appear here?

Indeed. Those paths are specified in this [PR](https://github.com/dfinity/interface-spec/pull/163) that could not be merged yet because the replica changes are still on hold due to performance benchmarks.

Author: > I would keep them in the spec as long as we keep them in the actual implementation. If we want to deprecate them (by forbidding the paths in the HTTP handler), then I'd announce this intention as a note in the spec (at the moment we're just saying that the subnet form "should" be used and we can strengthen this statement in a further note).

I'm fine with this, can you please edit that part in a way that it makes sense to you?

Reviewer: I've added a deprecation announcement and a note. Please close this thread if the formulation makes sense to you.
